### DISCUSSION TITLE
Find Endpoints of PolyData Curve

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#5017] [Q&A] [answered]
Is there a way to find the endpoints of a `PolyData` curve? Consider the case of performing an `intersection` between two meshes, returning an open 3D curve with 2 endpoints (a "start" and an "end"). How can one find these?

**Comment by darikg** [comment]
if `polydata.n_lines == 1`, you can read the vertex indices directly out of the line:

```python
import numpy as np
import pyvista as pv

line = pv.Line(resolution=10)
start_vertex = line.lines[1]  # line.lines[0] is the number of vertices in the line
end_vertex = line.lines[-1]
```

if the line is represented as a collection of edges, i.e. `len(polydata.lines) == polydata.n_lines * 3` we can look for vertices of degree 1. (Interior vertices in the line have degree 2.) It should be possible to do this somehow with the vtkVertexDegree filter, but that wants a vtkGraph, not a vtkPolyData and I don't know how to (easily) convert between them. Here's a kludgey way to do it:

```python
# Construct the line
n = 10
pts = np.random.uniform(-1, 1, (n, 3))
i = np.arange(n)
edges = np.stack([2 * np.ones(n - 1, dtype='int'), i[:-1], i[1:]], axis=1)  # Edges (0, 1), (1, 2), (2, 3), etc
line = pv.PolyData(pts, lines=edges)

# Find the vertex that starts an edge but doesn't end one, and vice versa
edges = line.lines.reshape(-1, 3)
a = set(edges[:, 1])  # Vertices that start an edge
b = set(edges[:, 2])  # Vertices that end an edge
start_vertex = (a - b).pop()  # Set difference, assume only one vertex in the result
end_vertex = (b - a).pop()
```

This won't work if we can't rely on edge direction, in which case we could simply count vertices:
```python
from collections import Counter
c = Counter(edges[:, 1])
c.update(edges[:, 2])
end_point_vertices = [i for (i, n) in c.items() if n == 1]
```

**Comment by edabor** [comment, marked as answer]
I guess that if you are sure that your `PolyData` only contains lines, you can use the connectivity to get the points that are only used by a single cell (which are the end points).
See the connectivity function https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.point_cell_ids.html#pyvista.DataSet.point_cell_ids
```python
import pyvista as pv

points = np.array([[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 1]])
line = pv.lines_from_points(points)

mask = [len(line.point_cell_ids(i)) == 1 for i in range(line.n_points)]
np.where(mask)
```
(array([0, 3]),)

**Reply by adamgranthendry** [reply]
Thank you, I should have looked at this more closely. This certainly works as the answer (caveat, yes, that the `PolyData` only contains lines).

**Comment by adamgranthendry** [comment]
> It should be possible to do this somehow with the vtkVertexDegree filter, but that wants a vtkGraph

This approach makes most sense to me (it is a standard graph theory problem). However, the points in my curve are unordered, so I don't know how to add edges a-priori.

In the example below, the `PolyData` object `curve` has edges inserted properly. Is there a way to get the edges each vertex in the `curve` is connected to? Then I could fix this line:

```python
edges = [graph.AddEdge(i, i+1) for i in range(curve.n_points - 1)]
```

## Example
```python
import pyvista as pv
import vtkmodules.all as vtk

hills = pv.ParametricRandomHills().extract_surface().triangulate()

xi, xf, yi, yf, zi, zf = hills.bounds

width = xf - xi
height = yf - yi

plane = pv.Plane(
    hills.center,
	direction=(1,1,1),
	i_size=1.25*width,
	j_size=1.25*height,
).triangulate().subdivide(4)

curve, _, _ = plane.intersection(hills, split_first=False, split_second=False)

graph = vtk.vtkMutableUndirectedGraph()

verts = [graph.AddVertex() for _ in range(curve.n_points)]
edges = [graph.AddEdge(i, i+1) for i in range(curve.n_points - 1)]

points = vtk.vtkPoints()
for id in range(curve.n_points):
    points.InsertNextPoint(curve.points[id])

graph.SetPoints(points)

layout = vtk.vtkGraphLayoutView()
layout.AddRepresentationFromInput(graph)
layout.SetLayoutStrategyToPassThrough()

layout.Render()

degree= vtk.vtkVertexDegree()
degree.SetInputData(graph)
degree.Update()

endpoints = []
for i in range(curve.n_points):
    if valence.GetOutput().GetDegree(i) == 1:
	    endpoints.append(i)

plotter = pv.Plotter()

plotter.add_mesh(hills, show_edges=True)
plotter.add_mesh(plane, show_edges=True, opacity=0.5, color='red')
plotter.add_mesh(curve, color='orange', line_width=3)

plotter.fly_to(hills.center)
plotter.view_isometric()

plotter.show()
layout.Render()
```

## Inputs
![image](https://github.com/pyvista/pyvista/assets/77850046/4e4ffd7e-7458-4fb0-b331-965ce9eba81b)

## Graph
![image](https://github.com/pyvista/pyvista/assets/77850046/d1bf243c-e5cb-4df8-8db3-6a65b49d3760)

**Comment by edabor** [comment]
> In the example below, the `PolyData` object `curve` has edges inserted properly. Is there a way to get the edges each vertex in the `curve` is connected to? Then I could fix this line:
> 
> ```python
> edges = [graph.AddEdge(i, i+1) for i in range(curve.n_points - 1)]
> ```

For each point, you can get the all the edges ids connected to this point with `point_cell_ids(i)`. See the documentation (the `0.42` documentation image is broken, so I took the `0.40`). https://docs.pyvista.org/version/0.40/api/core/_autosummary/pyvista.DataSet.point_cell_ids.html

Here is your example with the end-points labelled. Is this what you expected ? I don't know much of `vtkVertexDegree` so your approach might also work :smile: 
```python
import pyvista as pv
import vtkmodules.all as vtk

hills = pv.ParametricRandomHills().extract_surface().triangulate()

xi, xf, yi, yf, zi, zf = hills.bounds

width = xf - xi
height = yf - yi

plane = (
    pv.Plane(
        hills.center,
        direction=(1, 1, 1),
        i_size=1.25 * width,
        j_size=1.25 * height,
    )
    .triangulate()
    .subdivide(4)
)

curve, _, _ = plane.intersection(hills, split_first=False, split_second=False)
mask = [len(curve.point_cell_ids(i)) == 1 for i in range(curve.n_points)]
mask = np.where(mask)[0]

plotter = pv.Plotter()

plotter.add_mesh(hills, show_edges=True)
plotter.add_mesh(plane, show_edges=True, opacity=0.5, color='red')
plotter.add_mesh(curve, color='orange', line_width=3)
plotter.add_point_labels(curve.points[mask, :], labels=["end points"] * len(mask))

plotter.fly_to(hills.center)
plotter.view_isometric()

plotter.show()
```

**Reply by adamgranthendry** [reply]
Thank you, yes. I should have looked at your previous post in greater detail.

> I don't know much of vtkVertexDegree so your approach might also work 😄

You've acheived what I was trying to accomplish using that. The degree/valence of a vertex in a graph is the number of edges connected to it, which is precisely what we get by querying `len(curve.point_cell_ids(i))`.

My head was stuck in graph theory mode, so I was thinking of using VTK tools that referenced similar terms ("graph", "degree", etc.).

> Here is your example with the end-points labelled. Is this what you expected ?

Yes, and you can see from the output below that it works correctly:

![image](https://github.com/pyvista/pyvista/assets/77850046/06a97360-c7ce-4184-a43a-324bdb904753)

**Reply by adamgranthendry** [reply]
Some follow-ups (i.e. "recommendations"/"requests"):

1. The fact `point_cell_ids` works in this case is slightly confusing since my `PolyData` is a curve that technically has no cells. Is it possible to implement a method `point_edge_ids`? I'm sure users will want a way to find the edges connected to a vertex in a mesh.

2. A next step with `curve` is to turn it into a [Spline](https://docs.pyvista.org/version/stable/api/utilities/_autosummary/pyvista.Spline.html), which requires the points/nodes be ordered:

> Array must be 3D and directionally ordered.

 Now that I know the end points, is there a way to traverse the `PolyData` from one end point to the other to get the "proper ordering" of the nodes (i.e. think a DirectedGraph from one end point to the other, if this makes more sense intuitively).

**Reply by adamgranthendry** [reply]
> The fact point_cell_ids works in this case is slightly confusing since my PolyData is a curve that technically has no cells.

I take that back. I guess it _does_ actually have "cells":

```python
PolyData (0x15d964904c0)
  N Cells:    598
  N Points:   599
  N Strips:   0
  X Bounds:   -7.528e+00, 1.000e+01
  Y Bounds:   3.500e+00, 2.000e+01
  Z Bounds:   3.119e-01, 4.913e+00
  N Arrays:   5
```

I think "n-simplex", but the explanation is [here](https://docs.pyvista.org/version/0.40/user-guide/what-is-a-mesh.html#what-is-a-cell):

> Cells aren’t limited to voxels, they could be a triangle between three points, a line between two points, or even a single point could be its own cell (but that’s a special case).

Another "language-barrier" problem 😄

**Reply by darikg** [reply]
> In the example below, the PolyData object curve has edges inserted properly. Is there a way to get the edges each vertex in the curve is connected to? Then I could fix this line:
> `edges = [graph.AddEdge(i, i+1) for i in range(curve.n_points - 1)]`

Using @beroda's demo to get a curve:

```python
assert len(curve.lines) == curve.n_lines * 3
edges = curve.lines.reshape(-1, 3)[:, 1:]   # (n_edges, 2) array of vertex indices
edges = [graph.AddEdge(u, v) for (u, v) in edges]
```

libigl also has an [edges_to_path](https://libigl.github.io/libigl-python-bindings/igl_docs/#edges_to_path) method that will convert that (n_edges, 2) array to a vector of vertex indices (and some other convenient index arrays)

**Reply by edabor** [reply]
>    The fact `point_cell_ids` works in this case is slightly confusing since my `PolyData` is a curve that technically has no cells. Is it possible to implement a method `point_edge_ids`? I'm sure users will want a way to find the edges connected to a vertex in a mesh.

You can achieve with this small function:
```python
def get_edge_ids(curve: pv.PolyData, i: int):
    edge_ids = []
    for c in curve.point_cell_ids(i):
        if curve.get_cell(c).type is pv.CellType.LINE:
            edge_ids.append(c)
    return edge_ids
```

**Reply by adamgranthendry** [reply]
+1 for the 2 ways to assert the `PolyData` is indeed a curve:

```python
assert len(curve.lines) == curve.n_lines * 3
```

or

```python
assert all([curve.get_cell(c).type is pv.CellType.LINE for c in curve.point_cell_ids(i) for i in range(curve.n_points)])
```

> libigl also has an [edges_to_path](https://libigl.github.io/libigl-python-bindings/igl_docs/#edges_to_path) method that will convert that (n_edges, 2) array to a vector of vertex indices

Oooo, yes! This helps answer my [question #2](https://github.com/pyvista/pyvista/discussions/5017#discussioncomment-7210601).

After:

```
py -m pip install libigl
```

we modify the example to this, which shows the oriented path:

```python
import igl
import numpy as np
import pyvista as pv
import vtkmodules.all as vtk

plotter = pv.Plotter()

hills = pv.ParametricRandomHills().extract_surface().triangulate()

xi, xf, yi, yf, zi, zf = hills.bounds

width = xf - xi
height = yf - yi

plane = pv.Plane(
    hills.center,
	direction=(1,1,1),
	i_size=1.25*width,
	j_size=1.25*height,
).triangulate().subdivide(4)

curve, _, _ = plane.intersection(hills, split_first=False, split_second=False)

# Assert curve is actually a curve (no areas)
assert len(curve.lines) == curve.n_lines * 3
edges = curve.lines.reshape(-1,3)[:, 1:]  # (n_edges, 2) array of vertex indices

path, _, _ = igl.edges_to_path(edges)

hills.actor = plotter.add_mesh(hills, show_edges=True)
plane.actor = plotter.add_mesh(plane, show_edges=True, opacity=0.5, color='red')
curve.actor = plotter.add_mesh(curve, color='orange', line_width=3)

plotter.add_point_labels(
    curve.points[path],
	labels=[f'{ndx}' for ndx in range(curve.n_points)],
	text_color='white',
	font_size=20,
)

plotter.fly_to(hills.center)
plotter.view_isometric()

plotter.show()
```

## Ordered Points
![image](https://github.com/pyvista/pyvista/assets/77850046/64e7d8fe-6b5d-437d-848f-d13f908ad1de)

and now we can make a spline:

```python
spline = pv.Spline(curve.points[path])
spline.actor = plotter.add_mesh(spline, color='yellow', line_width=4)
```

# Spline
![image](https://github.com/pyvista/pyvista/assets/77850046/52e091ec-58b6-4bdb-8259-7379f5b222c6)

**Reply by darikg** [reply]
> +1 for the 2 ways to assert the PolyData is indeed a curve:
> `assert len(curve.lines) == curve.n_lines * 3`

One tiny clarification: this doesn't really check that -- it's just a quick sanity check to ensure it's reasonable to reshape `curve.lines` into a (n_lines, 3) array. It's still possible for the PolyData to contain non-line cells. What I wanted to distinguish there was the difference between   e.g. `lines = [2, 0, 1, 2, 1, 2]` (two 2-vertex lines (0, 1) and (1, 2)) and `lines = [3, 0, 1, 2]` (one 3-vertex line (0, 1, 2)).

**Reply by adam-grant-hendry** [reply]
Ah, I see; good to know. Thanks for the clarification!
